cdc/scheduler/internal/tp/scheduler_manager.go

```diff
@@ -0,0 +1,133 @@
+) *schedulerManager {
+	sm := &schedulerManager{
+		changefeedID: changefeedID,
+		schedulers:   make(map[schedulerType]scheduler),
+		tasksCounter: make(map[struct {
+			scheduler string
+			task      string
+		}]int),
+	}
+
+	sm.schedulers[schedulerTypeBasic] = newBasicScheduler()
+	sm.schedulers[schedulerTypeBalance] = newBalanceScheduler(balanceInterval)
+	sm.schedulers[schedulerTypeMoveTable] = newMoveTableScheduler()
+	sm.schedulers[schedulerTypeRebalance] = newRebalanceScheduler()
+	sm.schedulers[schedulerTypeDrainCapture] = newDrainCaptureScheduler()
+
+	return sm
+}
+
+func (sm *schedulerManager) Schedule(
+	checkpointTs model.Ts,
+	currentTables []model.TableID,
+	aliveCaptures map[model.CaptureID]*model.CaptureInfo,
+	replications map[model.TableID]*ReplicationSet) []*scheduleTask {
+
+	for _, scheduler := range sm.schedulers {
+		tasks := scheduler.Schedule(checkpointTs, currentTables, aliveCaptures, replications)
+		for _, t := range tasks {
+			name := struct {
+				scheduler, task string
+			}{scheduler: scheduler.Name(), task: t.Name()}
+			sm.tasksCounter[name]++
+		}
+		if len(tasks) != 0 {
+			log.Info("tpscheduler: new schedule task",
+				zap.Int("task", len(tasks)),
+				zap.String("scheduler", scheduler.Name()))
+			return tasks
+		}
+	}
+
+	return nil
+}
+
+func (sm *schedulerManager) MoveTable(tableID model.TableID, target model.CaptureID) {
+	scheduler, ok := sm.schedulers[schedulerTypeMoveTable]
+	if !ok {
+		log.Panic("tpscheduler: move table scheduler not found")
+	}
+	moveTableScheduler, ok := scheduler.(*moveTableScheduler)
+	if !ok {
+		log.Panic("tpscheduler: invalid move table scheduler found")
+	}
+	if !moveTableScheduler.addTask(tableID, target) {
+		log.Info("tpscheduler: manual move Table task ignored, "+
+			"since the last triggered task not finished",
+			zap.Int64("tableID", tableID),
+			zap.String("targetCapture", target))
+	}
+}
+
+func (sm *schedulerManager) CollectMetrics() {
+	cf := sm.changefeedID
+	for name, counter := range sm.tasksCounter {
+		scheduleTaskCounter.
+			WithLabelValues(cf.Namespace, cf.ID, name.scheduler, name.task).
+			Add(float64(counter))
+		sm.tasksCounter[name] = 0
+	}
+}
+
+func (sm *schedulerManager) Rebalance() {
+	scheduler, ok := sm.schedulers[schedulerTypeRebalance]
+	if !ok {
+		log.Panic("tpscheduler: rebalance scheduler not found")
+	}
+	rebalanceScheduler, ok := scheduler.(*rebalanceScheduler)
+	if !ok {
+		log.Panic("tpscheduler: invalid rebalance scheduler found")
+	}
+
+	atomic.StoreInt32(&rebalanceScheduler.rebalance, 1)
+}
+
+func (sm *schedulerManager) DrainCapture(target model.CaptureID) bool {
+	scheduler, ok := sm.schedulers[schedulerTypeDrainCapture]
+	if !ok {
+		log.Panic("tpscheduler: drain capture scheduler not found")
+	}
+	drainCaptureScheduler, ok := scheduler.(*drainCaptureScheduler)
+	if !ok {
+		log.Panic("tpscheduler: invalid drain capture scheduler found")
+	}
+
+	return drainCaptureScheduler.setTarget(target)
```

Comment:
Should it check if there is a capture is draining already?

Reply:
if some capture is draining, `setTarget` return false.